server/server_test.go

```diff
@@ -787,31 +791,105 @@ func TestCrossClientScopes(t *testing.T) {
 	if err := s.storage.CreateClient(peer); err != nil {
 		t.Fatalf("failed to create client: %v", err)
 	}
+	fixture.peer = &peer
 
-	oauth2Config = &oauth2.Config{
-		ClientID:     client.ID,
-		ClientSecret: client.Secret,
-		Endpoint:     p.Endpoint(),
-		Scopes: []string{
-			oidc.ScopeOpenID, "profile", "email",
-			"audience:server:client_id:" + client.ID,
-			"audience:server:client_id:" + peer.ID,
-		},
-		RedirectURL: redirectURL,
-	}
+	reqDump, respDump := crossClientsTest(&fixture, &testDeferred)
 
-	resp, err := http.Get(oauth2Server.URL + "/login")
-	if err != nil {
-		t.Fatalf("get failed: %v", err)
-	}
-	if reqDump, err = httputil.DumpRequest(resp.Request, false); err != nil {
-		t.Fatal(err)
-	}
-	if respDump, err = httputil.DumpResponse(resp, true); err != nil {
-		t.Fatal(err)
+	if !gotCode {
+		t.Errorf("never got a code in callback\n%s\n%s", reqDump, respDump)
 	}
 }
 
+type crossClientsFixture struct {
+	provider     *oidc.Provider
+	client       *storage.Client
+	peer         *storage.Client
+	oauth2Server *httptest.Server
+}
+
+type crossClientsTestDeferred struct {
```

Comment:
What are these new types? I'm confused by the naming here. Can you add some explanations or not refactor the tests?

Reply:
I introduced new types because there were cyclic dependencies in this code, and I wanted to somehow refactor it. But I will remove this refactoring as you suggested in overall comment, and I will just duplicate this test for the new case.